### DISCUSSION TITLE
`Nunit.Framework.Internal.Test`'s `Fixture` property is null from inside `ITestBuilder`

### DISCUSSION THREAD

**Discussion by awa5114** [nunit/nunit#4372] [Q&A] [answered]
As a follow up to a [previous issue](https://github.com/nunit/nunit/issues/4349) I have been trying to implement a custom attribute using the `ITestBuilder` interface as follows:

```
[AttributeUsage(AttributeTargets.Method)]
   internal class TestBuilderAttribute : Attribute, ITestBuilder
   {
      public IEnumerable<TestMethod> BuildFrom(IMethodInfo method, NUnit.Framework.Internal.Test suite)
      {
         throw new NotImplementedException();
      }
   }
```

I decorated a test case as follows:

```
[TestFixture]
   public class MyTestFixture
   {
      [SetUp]
      public void SetUp()
      {
         ...
      }

      [TestBuilder]
      [Test]
      public void MyTest1()
      {
         ...
      }

      [TestBuilder]
      [Test]
      public void MyTest2()
      {
         ...
      }

      public void PrintHello()
      {
         Console.WriteLine("hello world");
      }
```

I would like to access `PrintHello` such that it gets triggered before each test. Unfortunately, from inside the `BuildFrom()` method in `ITestBuilder`, `suite.Fixture` is null, so I cannot downcast it as indicated in the [previous issue](https://github.com/nunit/nunit/issues/4349).

Am I doing something wrong here? Why is `suite.Fixture` null-valued even though I clearly have a fixture?

**Comment by stevenaw** [comment]
Would ICommandWrapper work for you? There is a derived IWrapSetUpTeardown marker interface you can implement to inject code to run in the SetUp/Teardown phase.

https://docs.nunit.org/articles/nunit/extending-nunit/ICommandWrapper-Interface.html

The docs suggest using MaxTimeAttribute as a simple, minimal example to reference: https://github.com/nunit/nunit/blob/master/src/NUnitFramework/framework/Attributes/MaxTimeAttribute.cs

**Comment by stevenaw** [comment]
Oh sorry I see now a bit more from the original issue. It's difficult to know without seeing the code you have which isn't working. Are you able to share a bit of what you have tried to implement in your custom attribute?

**Comment by manfred-brands** [comment, marked as answer]
@amine-aboufirass The `suite` passed in to the `BuildFrom` method is the `Fixture` you are in and the property `ClassName` is its type. Note that you cannot call a non-static method as at the time the `ITestBuilder` is run, it is to create tests, not to run them and the class will not be instantiated.

I'm still confused what you are trying to do and why you cannot call your `PrintHello` from the `[SetUp]` method.

@stevenaw Indeed the derived `ICommandWrapper` allows you to wrap tests, but the needed class `TestCommand` is in the `Internal` namespace so I don't know if that is really meant to be used by outside code?

**Comment by stevenaw** [comment]
@manfred-brands Huh, you are quite right, the callbacks passed in there are for a `TestExecutionContext` in the internal namespace. I'm surprised as we also recommend `ICommandWrapper` as an extension point in https://docs.nunit.org/articles/nunit/extending-nunit/Custom-Attributes.html, which would require an implementer of a custom command to receive a `TestExecutionContext` in the constructor as well as to extend one of the `Command`s in the internal namespace. This seems a bit contradictory.

There's a bit of a discussion about intent of the types in the internal namespace here https://github.com/nunit/nunit/issues/1426, but as far as I can tell it wasn't clear-cut and still requires a more thorough cataloging and decision about the different types in the (sub)namespaces.

I think you are right, it would help to know a bit more about what @amine-aboufirass is trying to do.

**Comment by OsirisTerje** [comment]
This is more about usage, so moving it to Q&A discussions.
